### PR TITLE
Parse matrix-schemed URIs

### PR DESCRIPTION
**Pull request by turt2live**
This doesn't handle creation or making linkify do the right thing. **edit** now also include creation and linkify. see video:
A step towards https://github.com/vector-im/element-web/issues/16875

https://user-images.githubusercontent.com/16718859/150311674-90cbb693-aa93-4a40-8d86-6608a13826e9.mp4

**edit** (from @toger5):
~~This uses a hack to parse `matrix: ` as a token. It adds `matrix` as a custom protocol using `linkifyjs.registerCustomProtocol` and then uses the `PROTOCOL` token that matches with `https`, `http`, `ftp` ... `matrix`. This means that `https:u/toger5:matrix.org` is detected and highlighted as link. Our link handler should then still do the right thing though. But it would be better if linkify would not highlight it as a link. Since this is fixed in linkify version 4 it might be the most time efficient to wait until v4 and not try to implement the workaround:
https://github.com/Hypercontext/linkifyjs/issues/167~~

**UPDATE**
This is now using a newer version of the linkify package (v4 dev branch)
We can just use registerCustomProtocol now and the custom parser (the first approach I tried to make it work with v3) can get removed 5022748e39daeecf8e7e36723118c4f8e48bfeb3
**I don't have the bandwidth to maintain this, so someone else is welcome to take this over.**

---
Here's what your changelog entry will look like:

## ✨ Features
 * Parse matrix-schemed URIs ([\#7453](https://github.com/matrix-org/matrix-react-sdk/pull/7453)).

Preview: https://61e996f7e3f64158d028a103--matrix-react-sdk.netlify.app
⚠️ Do you trust the author of this PR? Maybe this build will steal your keys or give you malware. Exercise caution. Use test accounts.

### CODE REVIEW

**Comment by toger5** [comment]
I think I can do better with the hack:
> This uses a hack to parse matrix: as a token. It adds matrix as a custom protocol using linkifyjs.registerCustomProtocol and then uses the PROTOCOL token that matches with https, http, ftp ... matrix. This means that https:u/toger5:matrix.org is detected and highlighted as link. Our link handler should then still do the right thing though. But it would be better if linkify would not highlight it as a link. Since this is fixed in linkify version 4 it might be the most time efficient to wait until v4 and not try to implement the workaround:
Hypercontext/linkifyjs#167

this can still be reviewed but there is one bug that needs to be fixed:
 - `https://matrix.to/#/@toger5:matrix.org` gets interpreted as a matrix: link therefore the # is not accepted so the link only goes to the hastag
 - then `@toger5:matrix.org` gets caught by the other plugin and we have two links
 the result looks like so:
 
image

**Comment by toger5** [comment]
This is now using a newer version of the linkify package (v4 dev branch)
We can just use registerCustomProtocol now and the custom parser (the first approach I tried to make it work with v3) can get removed 5022748

**Comment by toger5** [comment]
This does not build until we change an import in the @matrix-org/linkifyjs package:
@matrix-org/linkifyjs/lib/linkify-string.js
```diff
- var linkifyjs = require('linkifyjs');
+ var linkifyjs = require('@matrix-org/linkifyjs');
```

File: src/utils/permalinks/MatrixSchemePermalinkConstructor.ts [hunk 1]
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2019 The Matrix.org Foundation C.I.C.
```
Comment (jryans) [inline src/utils/permalinks/MatrixSchemePermalinkConstructor.ts:2]: Assuming this is new code, should be a new date?

```suggestion
Copyright 2022 The Matrix.org Foundation C.I.C.
```

**Review by jryans — APPROVED** [review]
Looks great to me, thanks for working on this! 😄

**Pull request merged**